cloudferrylib/scheduler/state.py

```diff
@@ -0,0 +1,252 @@
+        current = task_key(self.redis.get(self.CURRENT_TASK_KEY))
+        return TaskState.from_dict(self.redis.hgetall(current))
+
+    def get_history(self):
+        all_task_ids = self.redis.lrange(self.TASKS, 0, -1)
+        pipe = self.redis.pipeline()
+        for task_id in all_task_ids:
+            pipe.hgetall(task_id)
+        history = pipe.execute()
+        return map(TaskState.from_dict, history)
+
+    def reset(self, task_id):
+        """Changes state of :task_id and all the following to not started"""
+        self.redis.set(self.CURRENT_TASK_KEY, task_key(task_id))
+
+    def cleanup(self):
+        """Removes all state-related data from persistent storage. In other words,
+        CloudFerry 'forgets' all the tasks it executed previously."""
+        all_tasks = self.get_history()
+
+        self.redis.delete(all_tasks, self.TASKS, self.CURRENT_TASK_KEY)
+
+    def already_started(self, task_id):
+        task_id = task_key(task_id)
+        return self.redis.hget(task_id, 'state') not in [TASK_NOT_STARTED,
+                                                         None]
+
+    def build_schedule(self, migration):
+        """
+        Fills DB with scenario information on start
+        """
+        # if scenario was started previously:
+        #    fast forward scenario to CURRENT_STATE
+        # else:
+        #    rewrite scenario in DB
+        history = self.get_history()
+        new_scenario = [task_data(task, i) for i, task in enumerate(migration)]
+
+        scenario_changed = (len(history) == 0 or len(new_scenario) == 0 or
+                            any([old != new for old, new in
+                                 itertools.izip(history, new_scenario)]))
+
+        migration.to_start()
+
+        if scenario_changed:
+            pipe = self.redis.pipeline()
+            pipe.delete(history, self.TASKS)
+            pipe.set(self.CURRENT_TASK_KEY, 0)
+            for i, task in enumerate(migration):
+                tk = task_key(i)
+                td = task_data(task, i, TASK_NOT_STARTED)
+                pipe.hmset(tk, td)
+                pipe.rpush(self.TASKS, tk)
+            pipe.execute()
+            migration.to_start()
+        else:
+            current_task = int(self.redis.get(self.CURRENT_TASK_KEY))
+            migration.fast_forward_to(current_task)
+
+
+class SignalHandler(observers.Observable):
+    def __init__(self, siglist=None, handler=None, **kwargs):
+        super(SignalHandler, self).__init__(**kwargs)
+        if siglist is None:
+            siglist = [signal.SIGTERM, signal.SIGINT]
+        if handler is None:
+            handler = self.system_signal_handler
+        for signum in siglist:
+            signal.signal(signum, handler)
+
+    def system_signal_handler(self, signum, frame):
+        LOG.info("Signal %d caught, setting current task state to INTERRUPTED",
+                 signum)
+        self.notify_observers(None, TASK_INTERRUPTED)
```

Comment:
I believe it gonna fail with "Nonetype is not callable", because TaskStateObserver has no handler on TASK_INTERRUPTED